[Security Solution] Save timeline as new

janmonschke opened the pull request:
## Summary

This is the second part of the work for fixing timeline saving issues as described in https://github.com/elastic/kibana/issues/165053. The first part, replacing auto-save with a dedicated save button was done in this PR: https://github.com/elastic/kibana/pull/169239.

In this PR we're adding the `Save as new` option to the timeline save modal. This allows users to create a copy of the timeline without changing the current timeline. This is especially helpful in a multi-user environment where there are higher chances for users causing conflicting changes.

This is how the feature works:

https://github.com/elastic/kibana/assets/68591/dcec9938-fb28-4c3e-b916-8589c6e28441

Notice how the changes to the date range are only applied to the second timeline. The initial timeline remains unchanged.

### Changes

- Added a new route `/api/timeline/_copy` that expects a timeline id and a timeline object. It will take the timeline object from the body and create a new timeline from it. Then it will fetch all related saved objects (notes, pinned events) for the timeline id in the body and create copies of those. These will be associated to the copied timeline.
- Creating a copy of the associated saved search of a timeline is not done on the server because there is currently no server-side API to do that (there is an issue in the backlog for adding that though: https://github.com/elastic/kibana/issues/170580).
  - This means that the copy is happening in the frontend before making the copy request. While this feels inconsistent, it's the only way we can do this at the moment.
- The `Save as new` switch is only shown for timelines that have been saved before.
- A lot of work was done to make the change-tracking of the embedded saved search work correctly. This required to split up setting and initializing the saved search. On top, we're now storing the saved search object in the local timeline store in order to pass it to the copy request.
- Various stability and type fixes in the timeline epic
- There's no API documentation for the copy route because OpenAPI doesn't allow to specify private endpoints. We really don't want users to use this endpoint yet until the saved search copy can be performed on the backend.

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

janmonschke committed:
add basic timeline clone route

janmonschke committed:
accept a timeline object in the request

this allows for overwriting fields that should not be saved in the original timeline

janmonschke committed:
use the original dates on cloned notes

janmonschke committed:
call clone timeline is flag is set

janmonschke committed:
fetch saved search on the backend

janmonschke committed:
remove return

janmonschke committed:
change expected request body

janmonschke committed:
clone and add to frontend

janmonschke committed:
Replace current timeline with the clone

janmonschke committed:
clone -> copy

janmonschke committed:
fix loading spinners

Saving saved search is called too often and its dispatch of `endTimelineSaving` interferes with the actual timeline save operation.

The loading spinner still works great now.

janmonschke committed:
create new saved search when saving as new

janmonschke committed:
keep copy of savedSearch in redux

janmonschke committed:
`fields.index.id` is not needed for the equality check anymore

janmonschke committed:
temporarily disable the save as new toggle

janmonschke committed:
remove unused import

janmonschke committed:
Add persistence comments

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
fix updateTimeline type error

janmonschke committed:
fix type errors

janmonschke committed:
manually save timeline's saved search

janmonschke committed:
use initializeSavedSearch to prevent incorrect "change" flagging

janmonschke committed:
simplify timeline saving

janmonschke committed:
harden action type checks

ActionTimeline did not actually suffice to represent the timeline acitons

janmonschke committed:
remove unused import

janmonschke committed:
remove unused parameter

janmonschke committed:
add unit test to check logic for when the save as new switch is shown

janmonschke committed:
add save as new task to cypress

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
add cypress test

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
add unit tests for update and initiallze saved search

janmonschke committed:
remove now unneeded test

index has been removed from search comparator

janmonschke committed:
fix actions test

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
remove saved search logic from the server

janmonschke committed:
unit test copyTimeline

janmonschke committed:
add openApi docs

janmonschke committed:
there's no way to mark an API as private in OpenAPI

janmonschke committed:
internal routes have a different versions

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
use correct version

janmonschke committed:
it's actually `statusCode` and not `status_code`

janmonschke committed:
it's actually status_code

janmonschke committed:
use status_code

janmonschke committed:
add copy timeline tests

janmonschke committed:
remove unneeded comments

janmonschke committed:
disable the save button when the timeline(template) is immutable

lgestc commented on x-pack/plugins/security_solution/common/api/timeline/copy_timeline/copy_timeline_route.ts:
```diff
@@ -0,0 +1,15 @@
+ * 2.0.
+ */
+
+import * as rt from 'io-ts';
```
we now have `zod` in our codebase if you feel like trying something new:)

janmonschke replied:
Yeah, hopefully next time :D

lgestc commented on x-pack/plugins/security_solution/public/common/components/discover_in_timeline/use_discover_in_timeline_actions.tsx:
```diff
@@ -177,67 +180,99 @@ export const useDiscoverInTimelineActions = (
-          onTitleDuplicate: () => {},
-          copyOnSave: !savedSearchId,
-        });
-
-        if (!response || !response.id) {
-          throw new Error('Unknown Error occured');
+        if (!timelineRef.current.savedSearch) {
+          dispatch(
+            timelineActions.initializeSavedSearch({
+              id: TimelineId.active,
+              savedSearch,
+            })
+          );
```
could early return approach possibly reduce the amount of nested conditions we have here?

lgestc commented on x-pack/plugins/security_solution/public/timelines/store/timeline/reducer.ts:
```diff
@@ -520,6 +522,26 @@ export const timelineReducer = reducerWithInitialState(initialTimelineState)
+  .case(initializeSavedSearch, (state, { id, savedSearch }) => ({
+    ...state,
+    timelineById: {
+      ...state.timelineById,
```
hey, check out `immer` - we have in our codebase. it is a nice way to avoid doing spreads like these:) just fyi

janmonschke replied:
I wish, haha. There's actually a refactoring planned in which we migrate the timeline redux code to redux toolkit which comes with immer. <3

lgestc reviewed (approved):
left some ideas, otherwise looks good :)

janmonschke committed:
make sure timeline requests don't crash on core errors

janmonschke committed:
Merge branch 'main' into clone-timeline

banderror commented on x-pack/plugins/security_solution/server/routes/index.ts:
```diff
@@ -132,6 +133,7 @@ export const initRoutes = (
   cleanDraftTimelinesRoute(router, config, security);
   deleteTimelinesRoute(router, config, security);
   persistFavoriteRoute(router, config, security);
+  copyTimelineRoute(router, config, security);
```
Nit: Can we extract all timeline route registrations to a single function that would be owned by the Investigations team? See `registerPrebuiltRulesRoutes`, `registerRuleManagementRoutes`, etc.

janmonschke replied:
@banderror Good idea: https://github.com/elastic/kibana/pull/171027/commits/0edb30403c76bcf126bed27c8044a741e4e92ca3

banderror reviewed (approved):
D&R LGTM

janmonschke committed:
move timeline routes to its own file to prevent codeowner requests in the future

janmonschke committed:
Merge branch 'main' into clone-timeline

PhilippeOberti commented:
I feel like the UI is a bit confusing: while on a lot of softwares you usually have a `Save` and a `Save As` button, here we have a single `Save` button and only in the modal we can choose to save as a new timeline.
I feel like for users who just want to regularly save their timeline as they're making changes, this will quickly become annoying. You have to click on the `Save` button, then move your mouse to click on the `Save` button again inside the modal...
For the users who want to save as a new timeline, it might not be super obvious that they have to click on the `Save` button, then use the toggle. This wouldn't be obvious, especially because the first time they would save a timeline, the toggle wouldn't even be displayed! Maybe we could have it always displayed, but it would be toggled and disabled for first time saves? That way the users would know it's there if they want to save as new timeline later...

Another issue I found is the interaction in the save modal isn't as intuitive as it could be: I think we should have the toggle to `save as new` at the very top. I also think the toggle should drive wether the `title` and `description` inputs are disabled or not. Currently, it is easy to get into a weird place where I save a timeline, then I make some modifications to it and want to save to a new timeline, so we change the name but don't switch the toggle and click save directly. The timeline title does get updated, but I see an error message in a pop up, and the timeline isn't actually saved...

https://github.com/elastic/kibana/assets/17276605/c7750e32-752f-4a06-9140-635baedd28f8

While I understand that this is an user error, I feel like we should protect ourselves against it and make sure that users can't change anything if the toggle is set to false. It will prevent future SDH due to user errors!

I would also consider (if it hasn't been already by UIUX) to have 2 buttons (`Save` and `Save As`), maybe just icon buttons. I feel like this has been pretty standard for decades now and people should understand what it means... It would simplify the flows quite a bit, both for the users and for us

PhilippeOberti commented:
Also another weird behavior, that might not be related to this PR, but if I hit `Discard timeline` why do we still show the unsaved timeline modal?

https://github.com/elastic/kibana/assets/17276605/5e344a7e-8d0e-45d3-98bc-22b3c25c5699

Maybe I'm misunderstanding what **discard** means? I would expect if I discard that it means I don't care about it and I would not see another modal later on to save it?

logeekal commented:
>Maybe I'm misunderstanding what discard means? I would expect if I discard that it means I don't care about it and I would not see another modal later on to save it?

Discard means user starts from sctach with empty timeline. We show `unsaved` on empty/new timeline by default.

logeekal commented:
@PhilippeOberti @janmonschke 

>I feel like for users who just want to regularly save their timeline as they're making changes, this will quickly become annoying. You have to click on the Save button, then move your mouse to click on the Save button again inside the modal...

Completely agree with this. In my opinion as well `Save` should be 1-click thing.

One thing @codearos  pointed out was that `Edit Workflow` ( editing name and timeline) should be different that `save workflow`. But currently, we have both of them with same experience and we have to show a dialog box during `Save`. I think it is a design issue which is worth bringing up with Paul. 

Although, I would say that it does not affect this PR.

PhilippeOberti commented:
> Although, I would say that it does not affect this PR.

It kinda does though... I don't think we would want to merge a behavior that we know is not ideal. It will very rapidly go to Serverless, so I'd say if there are changes we know we want to make, we should do them before merging this PR

PhilippeOberti commented:
> > Maybe I'm misunderstanding what discard means? I would expect if I discard that it means I don't care about it and I would not see another modal later on to save it?
> 
> Discard means user starts from sctach with empty timeline. We show `unsaved` on empty/new timeline by default.

@logeekal yeah that makes sense, though I find that `unsaved` on an empty timeline infuriating personally. I really don't understand why we do it. I feel like we should mark the timeline as unsaved only when any value changes...

logeekal commented:
> @logeekal yeah that makes sense, though I find that unsaved on an empty timeline infuriating personally. I really don't understand why we do it. I feel like we should mark the timeline as unsaved only when any value changes...

Well, one change proposed by design is that there should be No message when timeline is actually `Saved`. and only show when that timeline is `unsaved` which makes sense to me because every tool will show some symbol when a file/asset is unsaved. For eg. a `*` or `+`

PhilippeOberti reviewed (changes requested):
@janmonschke after a discussion with Paul and the team, it seems that we're in agreement that having a single `Save` button that opens a modal to save and save as isn't ideal, but it might be sufficient for now. It gets us closer to the behavior in Discover.

I think it's worth starting a discussion with the Discover and UIUX teams to see if we can come up with a better solution...

What I believe is needed before merging this PR though is fixing the issue mentioned in [this comment](https://github.com/elastic/kibana/pull/171027#issuecomment-1828587104) (see the video) where the timeline isn't actually saved and throws an error on the UI.
Steps to reproduce:
- open and save a timeline
- make some changes to that timeline
- click the save button, then in the modal, modify the title and click save in the bottom right corner WITHOUT toggling the new timeline toggle
The title in the timeline modal changes, but a pop up at the bottom right shows an error. Navigating to the timeline page shows that the title was not actually changed.

janmonschke committed:
fix persistent 409 when saving new timelines twice

Sadly, `endTimelineSaving` has side effects, therefore we cannot simply call it in this context. Only calling it in the error case though works.

janmonschke commented:
@PhilippeOberti 

> I think it's worth starting a discussion with the Discover and UIUX teams to see if we can come up with a better solution...

100% agreed! And thanks for the good discussion on here. 

> What I believe is needed before merging this PR though is fixing the issue mentioned in [this comment](https://github.com/elastic/kibana/pull/171027#issuecomment-1828587104) (see the video) where the timeline isn't actually saved and throws an error on the UI.

Fixed in https://github.com/elastic/kibana/pull/171027/commits/862d12c0c3edccae7f6787e70256763d1b95145f

janmonschke committed:
Merge branch 'main' into clone-timeline

PhilippeOberti reviewed (approved):
Desk tested and code LGTM, thanks for fixing the bug!

janmonschke committed:
fix cypress tests

janmonschke committed:
Merge branch 'main' into clone-timeline

janmonschke committed:
Merge branch 'main' into clone-timeline
